ibm/resource_ibm_kms_key_policy.go

```diff
@@ -0,0 +1,476 @@
+	if err != nil {
+		return diag.Errorf("Get Key failed with error: %s", err)
+	}
+	err = resourceHandlePolicies(context, d, kpAPI, meta, key_id)
+	if err != nil {
+		return diag.Errorf("Could not create policies: %s", err)
+	}
+	d.SetId(key.CRN)
+	return resourceIBMKmsKeyPolicyUpdate(context, d, meta)
+}
+
+func resourceIBMKmsKeyPolicyRead(context context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	kpAPI, err := meta.(ClientSession).keyManagementAPI()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	crn := d.Id()
+	crnData := strings.Split(crn, ":")
+	endpointType := crnData[3]
+	instanceID := crnData[len(crnData)-3]
+	keyid := crnData[len(crnData)-1]
+
+	var instanceType string
+	var hpcsEndpointURL string
+	log.Println("instancetype", instanceType)
+	if crnData[4] == "hs-crypto" {
+		instanceType = "hs-crypto"
+		hpcsEndpointAPI, err := meta.(ClientSession).HpcsEndpointAPI()
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		resp, err := hpcsEndpointAPI.Endpoint().GetAPIEndpoint(instanceID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		if endpointType == "public" {
+			hpcsEndpointURL = "https://" + resp.Kms.Public + "/api/v2/keys"
+		} else {
+			hpcsEndpointURL = "https://" + resp.Kms.Private + "/api/v2/keys"
+		}
+
+		u, err := url.Parse(hpcsEndpointURL)
+		if err != nil {
+			return diag.Errorf("Error Parsing hpcs EndpointURL")
+
+		}
+		kpAPI.URL = u
+	} else if crnData[4] == "kms" {
+		instanceType = "kms"
+		if endpointType == "private" {
+			if !strings.Contains(kpAPI.Config.BaseURL, "private") {
+				kmsEndpURL := strings.SplitAfter(kpAPI.Config.BaseURL, "https://")
+				if len(kmsEndpURL) == 2 {
+					kmsEndpointURL := kmsEndpURL[0] + "private." + kmsEndpURL[1]
+					u, err := url.Parse(kmsEndpointURL)
+					if err != nil {
+						return diag.Errorf("Error Parsing kms EndpointURL")
+					}
+					kpAPI.URL = u
+				} else {
+					return diag.Errorf("Error in Kms EndPoint URL ")
+				}
+			}
+		}
+	} else {
+		return diag.Errorf("Invalid or unsupported service Instance")
+	}
+
+	kpAPI.Config.InstanceID = instanceID
+	// keyid := d.Id()
+	key, err := kpAPI.GetKey(context, keyid)
+	if err != nil {
+		return diag.Errorf("Get Key failed with error: %s", err)
+	}
+
+	d.Set("instance_id", instanceID)
+	d.Set("key_id", keyid)
+	d.Set("endpoint_type", endpointType)
+	if key.Expiration != nil {
+		expiration := key.Expiration
+		d.Set("expiration_date", expiration.Format(time.RFC3339))
+	} else {
+		d.Set("expiration_date", "")
+	}
+	d.Set(ResourceName, key.Name)
+	d.Set(ResourceCRN, key.CRN)
+	state := key.State
+	d.Set(ResourceStatus, strconv.Itoa(state))
+	rcontroller, err := getBaseController(meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	id := key.ID
+	crn1 := strings.TrimSuffix(key.CRN, ":key:"+id)
+
+	d.Set(ResourceControllerURL, rcontroller+"/services/kms/"+url.QueryEscape(crn1)+"%3A%3A")
+
+	policies, err := kpAPI.GetPolicies(context, keyid)
+
+	if err != nil {
```

Comment:
Add condition for 404 inside if err != nil block
when response code is 404 Set d.SetId("") and return nil